### PR TITLE
Add support of GPU preemption, rebased.

### PR DESCRIPTION
**Pull request by wangliwei-intel**

### CODE REVIEW

**Comment by MichalMrozek** [comment]
Please remove all EventsDebugEnable related debug logging.
Actually it would be good if only place where logging happens is around exec call.

**Comment by MichalMrozek** [comment]
Please make sure that new test names are following Given When Then naming pattern.

**Comment by MichalMrozek** [comment]
Plase avoid magic "7" 
int havePreemption = 7;

**Comment by MichalMrozek** [comment]
Can you please re base the patch, as it looks like there are conflicts ?

**Comment by wangliwei-intel** [comment]
Rebase my patch against latest source code.

**Comment by wangliwei-intel** [comment]
Correct

File: runtime/os_interface/linux/drm_buffer_object.cpp [hunk 1]
```diff
@@ -129,6 +129,14 @@ int BufferObject::exec(uint32_t used, size_t startOffset, unsigned int flags, bo
     execbuf.flags = flags;
     execbuf.rsvd1 = drmContextId;
 
+#if 0
```
Comment (MichalMrozek) [inline runtime/os_interface/linux/drm_buffer_object.cpp]: please remove this code

File: runtime/os_interface/linux/drm_neo.cpp [hunk 2]
```diff
@@ -132,10 +132,11 @@ bool Drm::is48BitAddressRangeSupported() {
 }
 
 void Drm::checkPreemptionSupport() {
-#if defined(I915_PARAM_HAS_PREEMPTION)
+#if defined(I915_SCHEDULER_CAP_PREEMPTION)
     int value = 0;
-    auto ret = getParam(I915_PARAM_HAS_PREEMPTION, &value);
-    preemptionSupported = (ret == 0 && value == 1);
+    auto ret = getParamIoctl(I915_PARAM_HAS_SCHEDULER, &value);
+    UNRECOVERABLE_IF(ret != 0);
```
Comment (MichalMrozek) [inline runtime/os_interface/linux/drm_neo.cpp]: what about kernels that doesn't have I915_PARAM_HAS_SCHEDULER ?

File: runtime/os_interface/linux/drm_buffer_object.cpp [hunk 3]
```diff
@@ -129,6 +129,14 @@ int BufferObject::exec(uint32_t used, size_t startOffset, unsigned int flags, bo
...
+    if (this->drm->preemptionSupported) {
+        printDebugString(DebugManager.flags.PrintDebugMessages.get(), stdout, "Executing buffers in low priority mode, ctx_id = %d.\n", execbuf.rsvd1);
+    } else {
```
Comment (MichalMrozek) [inline runtime/os_interface/linux/drm_buffer_object.cpp]: preemption doesn't mean that command buffers are executed in low priority mode

File: runtime/command_queue/enqueue_common.h [hunk 4]
```diff
@@ -640,6 +640,8 @@ CompletionStamp CommandQueueHw<GfxFamily>::enqueueNonBlocked(
         gtpinNotifyPreFlushTask(this);
     }
 
+    DebugManager.log(DebugManager.flags.PrintDebugMessages.get(), "preemption = ", static_cast<int>(dispatchFlags.preemptionMode));
+    printDebugString(DebugManager.flags.PrintDebugMessages.get(), stdout, "preemption = %d.\n", static_cast<int>(dispatchFlags.preemptionMode));
```
Comment (MichalMrozek) [inline runtime/command_queue/enqueue_common.h]: Please stick to one logging method, either via DebugManager or printDebugString.
I preffer to do this via DebugManager

File: runtime/os_interface/linux/drm_buffer_object.cpp [hunk 5]
```diff
@@ -129,6 +129,14 @@ int BufferObject::exec(uint32_t used, size_t startOffset, unsigned int flags, bo
     execbuf.flags = flags;
     execbuf.rsvd1 = drmContextId;
 
+#if 0
+    if (this->drm->preemptionSupported) {
```
Comment (MichalMrozek) [inline runtime/os_interface/linux/drm_buffer_object.cpp]: combine both section to one logging call.

File: runtime/os_interface/linux/drm_neo.cpp (hunk 2 above)
Comment (wangliwei-intel) [reply runtime/os_interface/linux/drm_neo.cpp]: i915_drm.h:
#define I915_PARAM_HAS_SCHEDULER 41
#define I915_SCHEDULER_CAP_ENABLED (1ul << 0)
#define I915_SCHEDULER_CAP_PRIORITY (1ul << 1)
#define I915_SCHEDULER_CAP_PREEMPTION (1ul << 2)

This block of code is controlled by 'I915_SCHEDULER_CAP_PREEMPTION', according to the rules of kernel, if 'I915_SCHEDULER_CAP_PREEMPTION' is defined, the 'I915_PARAM_HAS_SCHEDULER' must be defined too.
If 'I915_PARAM_HAS_SCHEDULER' is not defined, then the total block won't execute and 'preemptionSupported' defaults to 'false'.

Comment (MichalMrozek) [reply runtime/os_interface/linux/drm_neo.cpp]: In Neo we are using a copy of i915_drm.h located in 
vpg-compute-neo\third_party\uapi\drm\i915_drm.h

So this cap will always be present, but when code is executed on kernels that doesn't have this cap IOCTL will fail.

We shouldn't check for defines presence.

Comment (wangliwei-intel) [reply runtime/os_interface/linux/drm_neo.cpp]: thank you, understood.

**Comment by MichalMrozek** [comment]
Thank you for your contribution.

**Pull request closed without merging**